test/serializer/abstract/QualifiedCall.js

```diff
@@ -0,0 +1,24 @@
+// add at runtime: global.bar = {x: 1};
+let bar = global.__abstract ? __makeSimple(__abstract('object', 'bar')) : {x: 1};
+
+let foo = global.__abstract ? __abstract('function', '(function() { return this.x; })') : function() { return this.x; };
+bar.foo = foo;
+x = bar.foo();
```

Comment:
Can you add an additional test that tests this but then removes the method before serialization?

```js
bar.foo = foo;
x = bar.foo();
delete bar.foo;
```
and
```js
bar.foo = foo;
x = bar.foo();
bar.foo = function() {};
```

Reply:
I suppose this works because it's an abstract object, but what the object is not abstract but the function is?

Reply:
Good question. I'll add your test cases and see.

Reply:
Hmm. I guess the short answer is: "Only pure functions may be abstract." Hence any function that uses its this argument is disqualified. 

We can possibly make an exception for calls where the this argument is an intrinsic, since such calls will get serialized in order, along with any modifications of their properties. Right now, this will take care of the second case. It turns out that the first case will not work because deleting properties of intrinsic objects is not tracked.  That seems like a separate issue to me.

Reply:
As long as they only read immutable properties of `this`, they're still pure. It's a very common pattern to export modules like this:

```js
var foo = {
  bar() {
    return x + 5;
  },
  baz(x) {
    return this.bar(x);
  }
};
```

For this to be pure, it's enough that `bar` is immutable.

You could solve this by always calling it using `fn.call(thisArg, arg0, arg1...)`. At least if the property is mutated/mutable.

Reply:
Unfortunately, it is far from simple to establish immutability and it is definitely out of scope for this pull request.

Reply:
I accepted this PR but it is a bit worrying that this simply gets the wrong behavior without a warning when this case is hit.

In a follow up, would it be worth while having a way to taint a property as assumed immutable? So that if it gets deleted or mutated after this callsite it's an error/warning?

Reply:
I'm adding an error if an unfrozen concrete object is passed to an abstract function.